### PR TITLE
contest APIs

### PR DESCRIPTION
**Pull request by ridhishjain**
#### APIs:
* *`createContest`*
**`POST /contests/` Admin only**: create a contest
* *`updateContest`*
**`POST /contests/:contest_id/update` Moderator only**: update contest
* *`addModerator`*
**`POST /contests/:contest_id/moderator` Moderator only**: Add moderator
* *`removeModerator`*
**`DELETE /contests/:contest_id/moderator` Moderator only**: remove moderator (creator can't be deleted)
* *`addGroup`*
**`POST /contests/:contest_id/:groups` Moderator only**: Add group
* *`removeGroup`*
**`DELETE /contests/:contest_id/groups` Moderator only**: remove group

### CODE REVIEW

File: server/schema/contests/updateContest.js [hunk 1]
```diff
@@ -46,12 +46,12 @@ module.exports = {
...
-      type: 'object',
-      format: 'date-time',
+      type: 'string',
+      format: 'dateTimeFormat',
     },
     end_time: {
-      type: 'object',
-      format: 'date-time',
+      type: 'string',
+      format: 'dateTimeFormat',
```
Comment (ridhishjain) [inline server/schema/contests/updateContest.js:54]: custom validator in `updateContest` schema also changed, same as `createContest` schema

File: server/schema/contests/index.js [hunk 2]
```diff
@@ -1,11 +1,21 @@
...
 const addModeratorSchema = require('./addModerator')
 const removeModeratorSchema = require('./removeModerator')
-const updateContestSchema = require('./updateContest')
+const addGroupSchema = require('./addGroup')
+const removeGroupSchema = require('./removeGroup')
+
+ajv.addFormat('dateTimeFormat', (dateTimeString) => {
+  const timestamp = new Date(dateTimeString)
+  return timestamp instanceof Object && typeof timestamp === 'object'
+})
```
Comment (ridhishjain) [inline server/schema/contests/index.js:12]: custom validator for date time strings

File: server/schema/contests/createContest.js [hunk 3]
```diff
@@ -15,12 +15,12 @@ const schema = {
...
-      type: 'object',
-      format: 'date-time',
+      type: 'string',
+      format: 'dateTimeFormat',
     },
     end_time: {
-      type: 'object',
-      format: 'date-time',
+      type: 'string',
+      format: 'dateTimeFormat',
```
Comment (ridhishjain) [inline server/schema/contests/createContest.js:23]: custom validator used

File: server/routes/index.js [hunk 4]
```diff
@@ -51,6 +51,24 @@ router.use('/notifications', notificationsRouter.getCreatorNotificationsRouter)
 
 router.use('/search', searchRouter.searchUsersRouter)
 
+router.use('/contests', contestsRouter.createContestRouter)
+router.use('/contests', contestsRouter.updateContestRouter)
+router.use('/contests', contestsRouter.addModeratorRouter)
+router.use('/contests', contestsRouter.removeModeratorRouter)
+router.use('/contests', contestsRouter.addGroupRouter)
+router.use('/contests', contestsRouter.removeGroupRouter)
```
Comment (ridhishjain) [inline server/routes/index.js:59]: these are made

File: server/schema/contests/index.js [hunk 5]
```diff
@@ -1,11 +1,21 @@
...
+const updateContestSchema = require('./updateContest')
 const addModeratorSchema = require('./addModerator')
 const removeModeratorSchema = require('./removeModerator')
-const updateContestSchema = require('./updateContest')
+const addGroupSchema = require('./addGroup')
+const removeGroupSchema = require('./removeGroup')
+
+ajv.addFormat('dateTimeFormat', (dateTimeString) => {
+  const timestamp = new Date(dateTimeString)
+  return timestamp instanceof Object && typeof timestamp === 'object'
```
Comment (cjchirag7) [inline server/schema/contests/index.js:11]: This validator is also accepting invalid strings like "a", "b", etc.
So, I suggest this modifying this as :
```suggestion
  return !isNaN(Date.parse(dateTimeString));
```

**Review by cjchirag7 — CHANGES_REQUESTED** [review]
Appears that this validator needs to modified.
Rest, Everthing looks great !

File: server/schema/contests/updateContest.js (hunk 1 above)
Comment (cjchirag7) [reply server/schema/contests/updateContest.js:54]: cool

File: server/schema/contests/index.js (hunk 5 above)
Comment (cjchirag7) [reply server/schema/contests/index.js:11]: You can try running this in console :
```
const timestamp = new Date("a")
console.log(timestamp instanceof Object && typeof timestamp === 'object');
```

Comment (cjchirag7) [reply server/schema/contests/index.js:11]: @NBNARADHYA , what is your review ?

Comment (NBNARADHYA) [reply server/schema/contests/index.js:11]:
```suggestion
  Object.prototype.toString.call(timestamp) === '[object Date]'
```

Comment (cjchirag7) [reply server/schema/contests/index.js:11]: return ?

Comment (cjchirag7) [reply server/schema/contests/index.js:11]: Try running this in console :
```
var timestamp = new Date("a")
console.log(Object.prototype.toString.call(timestamp) === '[object Date]');
```

Comment (cjchirag7) [reply server/schema/contests/index.js:11]: Actually, the prototype would same, even for invalid dates, i guess. Better to use Date.parse(), it is working properly.

**Review by cjchirag7 — APPROVED** [review]
Great!

File: server/models/contests/updateContest.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      about,
+      rules,
+      prizes,
+      confidential_questions: confidentialQuestions,
+    } = body
+    const startTime = new Date(start_time)
+    const endTime = new Date(end_time)
+    const { contest_id: contestId } = params
+    pool.query(
+      `SELECT * FROM contests_moderators WHERE contest_id=? AND moderator=?`,
```
Comment (NBNARADHYA) [inline server/models/contests/updateContest.js:30]: Again..middleware

File: server/routes/contests/addGroup.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * @return {String}
+ */
+function sumErrors(errArray) {
+  const cb = (a, b) => a + b.message + ', '
+  return errArray.reduce(cb, '')
+}
+
+router.post(
+  '/:contest_id/groups',
+  middleware.verifyUserAccessToken,
```
Comment (NBNARADHYA) [inline server/routes/contests/addGroup.js:20]: `verifyModerator` middleware could be used here

File: server/routes/contests/addModerator.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ * @return {String}
+ */
+function sumErrors(errArray) {
+  const cb = (a, b) => a + b.message + ', '
+  return errArray.reduce(cb, '')
+}
+
+router.post(
+  '/:contest_id/moderator',
+  middleware.verifyUserAccessToken,
```
Comment (NBNARADHYA) [inline server/routes/contests/addModerator.js:20]: `verifyModerator` middleware

File: server/routes/contests/removeGroup.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ * @return {String}
+ */
+function sumErrors(errArray) {
+  const cb = (a, b) => a + b.message + ', '
+  return errArray.reduce(cb, '')
+}
+
+router.delete(
+  '/:contest_id/groups',
+  middleware.verifyUserAccessToken,
```
Comment (NBNARADHYA) [inline server/routes/contests/removeGroup.js:20]: `verifyModerator` middleware

File: server/routes/contests/removeModerator.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+ * @return {String}
+ */
+function sumErrors(errArray) {
+  const cb = (a, b) => a + b.message + ', '
+  return errArray.reduce(cb, '')
+}
+
+router.delete(
+  '/:contest_id/moderator',
+  middleware.verifyUserAccessToken,
```
Comment (NBNARADHYA) [inline server/routes/contests/removeModerator.js:20]: `verifyModerator` middleware

File: server/routes/contests/updateContest.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+ * @return {String}
+ */
+function sumErrors(errArray) {
+  const cb = (a, b) => a + b.message + ', '
+  return errArray.reduce(cb, '')
+}
+
+router.post(
+  '/:contest_id/update',
+  middleware.verifyUserAccessToken,
```
Comment (NBNARADHYA) [inline server/routes/contests/updateContest.js:20]: `verifyModerator` middleware

**Comment by ridhishjain** [comment]
`verifyModerator` middlware for contests:
i'll do this in a new PR for these APIs

* *`createContest`*
**`POST /contests/` Admin only**: create a contest
* *`updateContest`*
**`POST /contests/:contest_id/update` Moderator only**: update contest
* *`addModerator`*
**`POST /contests/:contest_id/moderator` Moderator only**: Add moderator
* *`removeModerator`*
**`DELETE /contests/:contest_id/moderator` Moderator only**: remove moderator (creator can't be deleted)
* *`addGroup`*
**`POST /contests/:contest_id/:groups` Moderator only**: Add group
* *`removeGroup`*
**`DELETE /contests/:contest_id/groups` Moderator only**: remove group

and rest will be built accordingly

**Pull request merged**